File: 34/py/main.py
def insertion_sort[T](array: list[T], comparator: callable[[T, T], bool] | None = None) -> list[T]:
    comparator = comparator or (lambda a, b: a - b)
    for i in range(1, len(array)):
        insert_val = array[i]
        j = i - 1
        while j >= 0 and comparator(array[j], insert_val) > 0:
            array[j + 1] = array[j]
            j -= 1
        array[j + 1] = insert_val
    return array

File: 34/go/main.go
package main

func insertionSort[T comparable](array []T, comparator func(a, b T) int) []T {
	if comparator == nil {
		comparator = func(a, b T) int {
			aa := any(a).(int)
			bb := any(b).(int)
			return aa - bb
		}
	}
	for i := 1; i < len(array); i++ {
		insertVal := array[i]
		insertIdx := i
		for j := i - 1; j >= 0; j-- {
			if comparator(array[j], insertVal) > 0 {
				array[j+1] = array[j]
				insertIdx = j
			} else {
				break
			}
		}
		array[insertIdx] = insertVal
	}
	return array
}
